phpseclib/Net/SSH2.php

```diff
@@ -2725,7 +2761,7 @@ public function read(string $expect = '', int $mode = self::READ_SIMPLE, int $ch
             $channel = $this->get_interactive_channel();
         }
 
-        if (!$this->isInteractiveChannelOpen($channel)) {
+        if (!$this->isInteractiveChannelOpen($channel) && empty($this->channel_buffers[$channel])) {
```

Comment:
I had previously considered this change to ignore the connectivity requirement if buffers were available but decided against it. The reason being is that, in the READ_SIMPLE mode, connectivity will be required regardless. Once the buffers are drained, we're still executing within the `while` loop below, and back to needing a connection when we invoke `get_channel_packet`, or hitting disconnect exception.

We could confine the buffer emptiness check to the READ_NEXT mode for now, or rework the `while` loop to stop once the buffers are empty (and the channel is closed), but as-is, I think we're just changing the exception bound to be thrown with this change.

Reply:
In the unit test I added (`testReadingOfClosedChannel()` to tests/Functional/Net/SSH2Test.php) it's doing READ_SIMPLE and it's able to read from a closed buffer without issue.

So in the `while (true)` loop in `read()` there's this:

```php
            $response = $this->get_channel_packet($channel);
            if ($response === true) {
                return Strings::shift($this->interactiveBuffer, strlen($this->interactiveBuffer));
            }
```
$response is true because of this (from `get_channel_packet`):

```php
                $response = $this->get_binary_packet(true);
                if ($response === true && $this->is_timeout) {
                    if ($client_channel == self::CHANNEL_EXEC && !$this->request_pty) {
                        $this->close_channel($client_channel);
                    }
                    return true;
                }
```
Now that said, it does get a little janky when `$ssh->setTimeout(0)` is set. Like it outputs some of the output of `$ssh->write("ping 127.0.0.2\n", SSH2::CHANNEL_SHELL);` when it shouldn't ever output anything as it should essentially be caught in an infinite loop. This is because of of https://github.com/phpseclib/phpseclib/issues/1440#issuecomment-635970968 . I could undo it but idk I think what we really need are per channel timeout's instead of a global timeout but I think that'd be best for another PR.

Could you give some sample code that demonstrates the problem you were concerned about?

Reply:
@terrafrost Yes, I'll try to get you that sample code tomorrow (or let you know that I've been unable to reproduce). Sorry it's taking me a bit to get to it.

Reply:
So here's a unit test I've come up to illustrate my point.
```
public function testReadFromChannelBufferWhenDisconnected(): void
    {
        $ssh = $this->getSSH2Login();
        $ssh->openShell();
        $sshChannel = $ssh->getInteractiveChannelId();
        // start a command that will run for a bit and then close the channel
        $ssh->write("for i in {1..2}; do echo \"sleep 2...\" && sleep 2; done; exit\n", $sshChannel);

        $ssh->setTimeout(5);
        $ssh->enablePTY();
        $ssh->exec('bash');
        $execChannel = $ssh->getInteractiveChannelId();
        $ssh->write('echo "hello"', $execChannel);
        $ssh->read('', SSH2::READ_SIMPLE, $execChannel);
        $ssh->disablePTY();
        // simulate server disconnect and generate error on read()
        $ssh->disconnect();

        // make channel_buffers public and uncomment
        // $this->assertNotEmpty($ssh->channel_buffers[$sshChannel], 'Buffered data on SSH channel was expected');
        $this->assertFalse($ssh->isInteractiveChannelOpen($sshChannel));

        $ssh->setTimeout(1);
        $readOutput = $ssh->read('', SSH2::READ_SIMPLE, $sshChannel);
        $this->assertStringContainsString('sleep 2...', $readOutput, 'Shell echo output was not found');
        $this->assertFalse($ssh->isInteractiveChannelOpen($sshChannel));
    }
```
Against master, this generates the error `phpseclib3\Exception\InsufficientSetupException : Operation disallowed prior to login()` whereas the error `phpseclib3\Exception\InvalidArgumentException : fsock is not a resource.` is generated by the proposed change to pivot in `read()` based on channel buffer data.

So in the case that the server has disconnected, we're still throwing an exception, and a less descriptive one at that with the change. However, that's not to suggest that `Operation disallowed prior to login()` is a good error for this case, and I actually strongly dislike that `read()` and `write()` will attempt to auto re-open a shell which has exited (and which can actually succeed if the connection is alive). I would rather receive the error `'Data is not available on channel'` like we do for other channel types in the case that the shell is no longer open, and which at least allows for acknowledgement that the prior session is gone. That change could be disruptive however, but perhaps it's not too bad so long as we maintain the shell auto-open (and not re-open) on `read()` and `write()`.

There is another interesting caveat to the proposed change, and which is highlighted by this test case as well. If removing the call to `$ssh->disconnect()`, the change does allow for continued read against the closed shell (as intended), albeit only if the connection remains alive, whereas master throws the error `phpseclib3\Exception\RuntimeException : Unable to fulfill channel request` in a failed attempt to re-open the shell (probably worthy of investigation as well, but I didn't pursue). I'm not sure how likely a scenario it is to maintain an active SSH connection with unexpected closure of the shell though, but it strikes me as rare enough that I wouldn't introduce general logic to support it.

So in summary, I think where I'm at is that I do like enabling access to the buffered channel data, and that `read()` is probably the appropriate place to grant it, but that we haven't fully arrived at that goal with the proposed change. Technically, the access is already there through `$mode=READ_NEXT`, though the shell re-open behavior does make that hard to work with.

Reply:
https://github.com/terrafrost/phpseclib/commit/f664ccb521819c075f727ffa9953cb4e7050e2e9 should take care of this.

And I do agree that the current behavior is suboptimal but, well, we've already talked about my long term plans lol. But right now my more immediate goals are completely rewriting ASN.1 / X.509. Like with what I have right now processing of ASN.1 stuff is deferred (ie. done on demand). So like if you want to get the total count of how many certs are in a CRL you don't need to parse the whole CRL - you can only parse the relevant parts of it. On a CRL with 40,000 entries this knocked the time down from 4s to 0.2s for that specific operation. I got other changes implemented, as well, but discussing them is a bit off topic in a PR for SSH2 lol